### ISSUE TITLE
Run hook before ORM loads

### ISSUE THREAD

**Issue by richdunajewski** [balderdashy/sails#3518]
I'm a little confused if this is possible, but let me give my scenario:

My DB configuration needs to be fetched from an external source during the bootstrap of Sails. I've created a custom hook that reaches out to the source and pulls in the config information and attempts to override the sails.config.connections objects where appropriate. The issue is that the ORM is already attempting to load before I can override the connection config, so it fails.

I've tried a couple different ways to do this:

1. Push the config logic into its own hook with no dependencies so it wouldn't need to wait for `hook:orm:loaded` to fire. This didn't seem to make a difference, as the ORM was still firing up connections before my code could intervene.

2. Run the code in the hook's `configure` function. The problem here is that much of the sails.* variables aren't available yet, so it's difficult to do what I need while using the framework (the bulk of the code is in a Service).

Hoping I'm missing something simple here. I almost need the reverse logic where the ORM is reliant on the custom hook rather than the other way around.

**Comment by japel** [comment]
@richdunajewski would you mind posting your hook (or something similar)...
nevertheless, these are the hooks loaded by sails: https://github.com/balderdashy/sails/blob/master/lib/app/configuration/default-hooks.js

@mikermcneil told me something about it:
> [...] So the hook:foo:loaded event gets fired when the callback of a hook's initialize function is fired
> Usually what you'll want to do is something like:
```javascript
// In your hook's initialize function...
sails.after('hook:foo:loaded', function (){
  // This will force your hook to finish doing all of its other hook-ish things
  // like binding the shadow routes in `routes.before` / `routes.after`, etc.
  // If you need to do lower level stuff when your hook loads, such as binding
  // shadow routes dynamically based on config, etc, then you'd do that in here
  // before calling cb().
  return cb();
});
```
> sails.after works basically like $(document).ready() in jQuery
> if that event has fired already, it will immediately trigger the callback (rather than what .on() would do-- i.e. listening for the next time it fires, but not calling the callback)
> you can also pass an array of event names into sails.after() and it will wait for all of them

so, have you tried using
sails.after('hook:request:loaded', function (){ ... })
?

**Comment by richdunajewski** [comment]
I think you mean `sails.after('hook:orm:loaded'. function(){})`, but yes I tried that and even your suggestion to wait until after request is loaded. The problem is that unless I can tell the ORM that it should depend on my hook, there's no way to alter the configuration before the ORM starts. Even if I can wedge my code in there, it would be a race condition and would be very unreliable.

If sails had a sails.before(), I could tell my hook to run a callback before the ORM is initialized, and that would alter the flow.

I don't think there's a mechanism that allows me to do this at this time. Perhaps this will explain my dilemma better:

1. `sails lift`
2. Create a native (not Waterline) DB connection to my configuration DB
3. Query for the rest of the configuration parameters that other adapters will need (such as Mongo, Redis, SQL)
4. Alter the sails.config.connections object with the result of the query in Step 3.
5. ORM loads as part of the `sails lift`, using the new/overridden parameters obtained in Step 4.
6. Lift completes

So in essence, one DB contains the configuration of the other adapters. I recognize this isn't ideal, but that's the current design to prevent everything from being in a config file in Sails. We need to run code before the ORM loads, and since it's a DB query, it's asynchronous. I can get code to run before the ORM registers connections, but the async nature doesn't guarantee that I can safely finish my query before the ORM is initialized.

**Comment by mikermcneil** [comment]
Hey @richdunajewski, thanks for providing such detailed info about your use case-- it helps a lot.

> My DB configuration needs to be fetched from an external source during the bootstrap of Sails

If I understand correctly, it sounds like your Sails server needs to load its `connections` configuration dynamically-- but only once right?  If you need to use a different database on a per-request basis, that's [a different story with a separate solution](http://stackoverflow.com/a/34845674/486547).

Secondly, just want to state my second assumption: that this is for use in production rather than dev (e.g. for white-label deployments where every customer gets their own Sails deployment with its own database). 

Assuming we're on the same page, here's what I would suggest:

Change your `app.js` file to have it send an HTTP request (e.g. with something like [request](https://github.com/request/request)) to get your connections config before you call `sails.lift()` programmatically.  Then pass in the connections configuration (and any other dynamic configuration you need) directly as options.  For example:

```javascript
// In your app.js file:

/**
 * app.js
 *
 * Use `app.js` to run your app without `sails lift`.
 * To start the server, run: `node app.js`.
 *
 * This is handy in situations where the sails CLI is not relevant or useful.
 *
 * For example:
 *   => `node app.js`
 *   => `forever start app.js`
 *   => `node debug app.js`
 *   => `modulus deploy`
 *   => `heroku scale`
 *
 *
 * The same command-line arguments are supported, e.g.:
 * `node app.js --silent --port=80 --prod`
 */

// Ensure we're in the project directory, so relative paths work as expected
// no matter where we actually lift from.
process.chdir(__dirname);



// Hard-coded customer id
// (could also expect an env variable-- e.g. to access it: `process.env.customer_id`)
var CUSTOMER_ID = 'walmart';


// Send an HTTP request and receive the response.
//
// NOTE:
// See `http://node-machine.org/machinepack-http/send-http-request` for docs of what
// comes back and how to use this. I'm just using this HTTP request method because
// i'm most familiar w/ it. Use whatever makes sense to you / you're familiar with.
var Http = require('machinepack-http');
Http.sendHttpRequest({
  url: '/dbconfig',
  baseUrl: 'https://my-secret-private-api-with-all-the-credentials',
  method: 'get',
  params: {
    customerId: CUSTOMER_ID
  }
}).exec({
  // An unexpected error occurred.
  error: function (err){
    // Fail and crash the process.
    throw err;
  },

  // OK.
  success: function (response){
    
    // ok it seemed to work, so let's grab the data out of there and now
    // we'll pass this stuff in as connections config to sails.lift()
    var dbConfig = JSON.parse(response.body);
      
     

    // Ensure a "sails" can be located:
    (function() {
      var sails;
      try {
        sails = require('sails');
      } catch (e) {
        console.error('To run an app using `node app.js`, you usually need to have a version of `sails` installed in the same directory as your app.');
        console.error('To do that, run `npm install sails`');
        console.error('');
        console.error('Alternatively, if you have sails installed globally (i.e. you did `npm install -g sails`), you can use `sails lift`.');
        console.error('When you run `sails lift`, your app will still use a local `./node_modules/sails` dependency if it exists,');
        console.error('but if it doesn\'t, the app will run with the global sails instead!');
        return;
      }

      // Try to get `rc` dependency
      var rc;
      try {
        rc = require('rc');
      } catch (e0) {
        try {
          rc = require('sails/node_modules/rc');
        } catch (e1) {
          console.error('Could not find dependency: `rc`.');
          console.error('Your `.sailsrc` file(s) will be ignored.');
          console.error('To resolve this, run:');
          console.error('npm install rc --save');
          rc = function () { return {}; };
        }
      }

      var _ = require('lodash');
      var configOverrides = _.extend({
        // Set up a `customerDb` datastore configuration
        connections: {
          customerDb: dbConfig
        },
        // Set this as the default connection
        models: {
          connection: 'customerDb'
        }
      }, rc('sails'));
        

      // Start server
      sails.lift(config);
    })();   
   
  },
});
```

Sorry the example is messy, short on time.  But that should be just about it!  Just keep in mind you'll need to start the server using `node app` or `npm start`  instead of `sails lift`, since you're passing in the config to sails programmatically, and `sails lift` won't run `app.js`.





(btw thanks for the help, @japel!)

**Comment by sailsbot** [comment]
Thanks for posting, @richdunajewski.  I'm a repo bot-- nice to meet you!

The issue queue in this repo is for verified bugs with documented features.  Unfortunately, we can't leave some other types of issues marked as "open".  This includes bug reports about undocumented features or unofficial plugins, as well as feature requests, questions, and commentary about the core framework. Please review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) for details.

If you're here looking for help and can't find it, visit [StackOverflow](http://stackoverflow.com), our [Google Group](https://groups.google.com/forum/#!forum/sailsjs) or our [chat room](https://gitter.im/balderdashy/sails?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge).  But please do not let this disrupt future conversation in this issue! I am only marking it as "closed" to keep organized.

Thanks so much for your help!

> If I am mistaken, and this issue is about a _critical bug with a documented feature_, please accept my sincerest apologies-- I am but a humble bot working to make GitHub a better place.  If you open a new issue, a member of the core team will take a look ASAP.

**Comment by richdunajewski** [comment]
@mikermcneil, thanks! This is what I was looking for. I was able to connect to the DB before `sails.lift()` but wasn't sure how to pass the config into the lift method without having Sails overwrite everything. Your example made it clear that I just need to extend the object that `rc('sails')` digs up.

I wasn't using customer-specific configurations, but the example was close enough. A central DB controls the connection info for each DB. If you make a change to the central DB and restart the application, it should read in the new value and connect to a different DB host, for example. That way there are no config files to change and only one DB connection is actually coupled to Sails -- the rest being dynamic.

:+1:

**Comment by emahuni** [comment]
I figured it out: I changed the name from `sails-hook-polymorphic-orm` to `sails-hook-et-polymorphic-orm`. The reason was that this hook was being loaded before the orm when installed from npm. In tests and `api/hooks` etc it is loaded well before `orm`. So changing the name to start with an `e` makes sure it is loaded before `orm` that start with an `o` and not after. This works but is very brittle. If for some reason the implementation behind loading of hooks changes that will break with it.

All this is because of the lack of `sails.before` method or any other way to load hooks in a particular order. the only way is to rename the hook to have preceedence alphabetically.
